src/mlpack/core/data/load_image_impl.hpp

```diff
@@ -19,8 +19,6 @@
 namespace mlpack {
 namespace data {
 
-#ifdef HAS_STB // Compile this only if stb is present.

```

Comment:
Hey @PranavReddyP16 Thanks for contribution. Can you please explain just by removing `#ifdef HAS_STB` is helping here. Because the code present in this file still depends upon `HAS_STB` . In this we are trying to remove depedency upon HAS_STB so that we can use this command https://github.com/mlpack/models/issues/35#issuecomment-572470604 without using -DHAS_STB flag.

Reply:
Yes, so now the code in this file does not need the HAS_STB macro to be defined for it to work.
Now if we want to load an image, it has nothing to do with the HAS_STB macro.
Was this not the expected output?

Also I have taken all the related code for image loading from this file and placed it in load_image.cpp
Please correct me if I'm wrong.

Reply:
Hey @PranavReddyP16 If you are able to load an image using your build like using this program.
```
  #include <mlpack/core.hpp>
  #define STB_IMAGE_IMPLEMENTATION
  #include <mlpack/core/data/load_image_impl.hpp>

  using namespace mlpack;
  using namespace mlpack::ann;
  using namespace arma;

  int main()
  {
      std::cout << "Reading data ..." << endl;
      arma::Mat<unsigned char> matrix;
      data::ImageInfo info;
      data::Load("saitama.png", matrix, info, false, true);
      std::cout<<matrix;
  }

```
and compiling this program using  command like this
```
g++ image.cpp -o cnn `pkg-config --cflags --libs mlpack armadillo` -fopenmp
```
without using `-DHAS_STB` flag. Then it is good to go otherwise try something else.

Reply:
Hey @himanshupathak21061998 I'm getting this error when I try to compile with the command you mentioned above: 

Package mlpack was not found in the pkg-config search path.
Perhaps you should add the directory containing `mlpack.pc'
to the PKG_CONFIG_PATH environment variable
Package 'mlpack', required by 'virtual:world', not found
/usr/bin/ld: /tmp/cccxXTii.o: in function `main':
test.cpp:(.text+0x1f7ae): undefined reference to `bool mlpack::data::Load<unsigned char>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, arma::Mat<unsigned char>&, mlpack::data::ImageInfo&, bool, bool)'
collect2: error: ld returned 1 exit status

Any help is appreciated

Reply:
@PranavReddyP16 Where do you compile it inside build directory or not?

Reply:
@himanshupathak21061998 why does `STB_IMAGE_IMPLEMENTATION` need to be `#define`d in your example program?  We should be able to just include `<mlpack/core.hpp>` and `<mlpack/core/data/load_image.hpp>` (if `core.hpp` doesn't already include that) and then your example code should work fine.

@PranavReddyP16 the goal is to make the code work as it currently does, but with any instances of the `HAS_STB` macro (which will still be needed) occuring *only* in the .cpp file.

The program that @himanshupathak21061998 should compile also with the `-DHAS_STB` flag.  When it's compiled with `-DHAS_STB`, it should work and print a matrix; when it's compiled without `-DHAS_STB`, then it should throw an exception when it's run.

Reply:
@rcurtin Sorry, I thought we are trying to change behaviour by refactoring. I removed #define from example.

Reply:
:+1: it *sort* of changes behavior of the codebase, in that it's no longer necessary to specify `HAS_STB` when compiling.  Also, I realize there is an error I wrote about the program's behavior.  I've edited my comment now.